modules/mono/editor/GodotTools/GodotTools/Export/ExportPlugin.cs

```diff
@@ -111,62 +113,78 @@ private void _ExportBeginImpl(string[] features, bool isDebug, string path, long
+            else if (features.Contains("universal"))
             {
-                Directory.CreateDirectory(Path.Combine(outputDataDir, dir.Substring(publishOutputTempDir.Length + 1)));
+                if (platform == OS.Platforms.MacOS)
+                {
+                    archs.Add("x86_64");
+                    archs.Add("arm64");
+                }
             }
 
-            foreach (string file in Directory.GetFiles(publishOutputTempDir, "*", SearchOption.AllDirectories))
+            foreach (var arch in archs)
             {
-                File.Copy(file, Path.Combine(outputDataDir, file.Substring(publishOutputTempDir.Length + 1)));
+                string ridOS = DetermineRuntimeIdentifierOS(platform);
+                string ridArch = DetermineRuntimeIdentifierArch(arch);
+                string runtimeIdentifier = $"{ridOS}-{ridArch}";
+                string projectDataDirName = $"{DetermineDataDirNameForProject()}_{arch}";
+                if (platform == OS.Platforms.MacOS)
+                {
+                    projectDataDirName = Path.Combine("Contents", "Resources", projectDataDirName);
+                }
+
+                // Create temporary publish output directory
+
+                string publishOutputTempDir = Path.Combine(Path.GetTempPath(), "godot-publish-dotnet",
+                    $"{Process.GetCurrentProcess().Id}-{buildConfig}-{runtimeIdentifier}");
+
+                _tempFolders.Add(publishOutputTempDir);
+
+                if (!Directory.Exists(publishOutputTempDir))
+                    Directory.CreateDirectory(publishOutputTempDir);
+
+                // Execute dotnet publish
+
+                if (!BuildManager.PublishProjectBlocking(buildConfig, platform,
+                        runtimeIdentifier, publishOutputTempDir))
+                {
+                    throw new InvalidOperationException("Failed to build project.");
+                }
+
+                string soExt = ridOS switch
+                {
+                    OS.DotNetOS.Win or OS.DotNetOS.Win10 => "dll",
+                    OS.DotNetOS.OSX or OS.DotNetOS.iOS => "dylib",
+                    _ => "so"
+                };
+
+                if (!File.Exists(Path.Combine(publishOutputTempDir, $"{GodotSharpDirs.ProjectAssemblyName}.dll"))
+                    // NativeAOT shared library output
+                    && !File.Exists(Path.Combine(publishOutputTempDir, $"{GodotSharpDirs.ProjectAssemblyName}.{soExt}")))
+                {
+                    throw new NotSupportedException(
+                        "Publish succeeded but project assembly not found in the output directory");
+                }
+
+                // Add to the exported project shared object list.
+
+                foreach (string file in Directory.GetFiles(publishOutputTempDir, "*", SearchOption.AllDirectories))
+                {
+                    AddSharedObject(file, tags: null, projectDataDirName);
+                }
```

Comment:
How does this part work? Will it copy all files to `projectDataDirName`? Or only shared objects?

Reply:
It will copy all files, ensure it's copied after extracting template, and sign native libs / executables.